### PR TITLE
Editor+

### PR DESCRIPTION
**Pull request by Rimmuru**
feat(world): add train
feat(train): add hijack train
feat(train): add train speed
feat(train): add delete train
feat(train): add derail train
feat(train): add drive train
feat(train): add exit train
feat(Troll): add set bounty cmd

### CODE REVIEW

**Comment by ghost** [comment]
Removing the 3rd argument on cayo teleport will force the player instead of just sending a request like it is now..

**Comment by Rimmuru** [comment]
it forces them as is

**Comment by ghost** [comment]
> it forces them as is

Look in game scripts. It doesn't. Feel free to lmk, it's been a while. Doesn't matter.

**Comment by Rimmuru** [comment]
> > it forces them as is
> 
> Look in game scripts. It doesn't. Feel free to lmk, it's been a while. Doesn't matter.

i tested both versions. they both force. i just took the params out of my script but yeah you're right about the params in the freemode script, i just checked. enjoy your xmas :)

**Comment by Loko415** [comment]
bounty does not work on other players but it does work on yourself

**Comment by urmommine** [comment]
Sorry maybe this is stupid request, can u add teleport all to vapucci beach idk its look fun

**Comment by Rimmuru** [comment]
> Sorry maybe this is stupid request, can u add teleport all to vapucci beach idk its look fun

easily, but i doubt that yim would want such an abusive feature

**Comment by ghost** [comment]
> > Sorry maybe this is stupid request, can u add teleport all to vapucci beach idk its look fun
> 
> 
> 
> easily, but i doubt that yim would want such an abusive feature

There is already a kick all option. Be as toxic as you want. But no crashes or money options

**Comment by urmommine** [comment]
> > Sorry maybe this is stupid request, can u add teleport all to vapucci beach idk its look fun
> 
> easily, but i doubt that yim would want such an abusive feature

Isn't the same like tp all to apartment or warehouses?

**Comment by Loko415** [comment]
fix bounty. Doesnt work.

File: src/util/troll.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+	inline void set_bounty_on_player(player_ptr target)
+	{
+		const size_t arg_count = 21;
+		int64_t args[arg_count] =
+		{
+            (int64_t)eRemoteEvent::Bounty,    
+            target->id(),
+            1,
+            10000,
```
Comment (Yimura) [inline src/util/troll.hpp]: 10k bounty is a modder only bounty value, allow for setting a custom value possibly.

Also weird indents.

**Review by Yimura — CHANGES_REQUESTED** [review]

**Comment by maybegreat48** [comment]
So like I've mentioned in the review, YimMenu already has most of the features you are trying to add. The only new thing added is bounty

**Comment by Rimmuru** [comment]
> So like I've mentioned in the review, YimMenu already has most of the features you are trying to add. The only new thing added is bounty

buddy what? no it doesnt.

File: src/util/troll.hpp (hunk 1 above)
Comment (Rimmuru) [reply src/util/troll.hpp]: done the value btws, the indents is the default vs tab indent

**Comment by maybegreat48** [comment]
[Teleport To Apartment](https://github.com/YimMenu/YimMenu/blob/7b3d4cb17dceb448b87237dfe69d78fc8bb5f4b9/src/views/network/view_session.cpp#L244) [Teleport To Island](https://github.com/YimMenu/YimMenu/blob/master/src/backend/commands/player/toxic/send_to_island.cpp) (I may have forgotten to add the UI for it lol)

**Review by Yimura — DISMISSED** [review]
Looks good, can be merged into `master`.

Edit:
nvm, just saw the comment from @maybegreat48

File: src/util/troll.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+#include "gta/enums.hpp"
+#include "services/players/player_service.hpp"
+#include "script_global.hpp"
+#include "pointers.hpp"
+
+namespace big::troll
+{
+	inline void set_bounty_on_player(player_ptr target, int value)
```
Comment (tupoy-ya) [inline src/util/troll.hpp:9]: Would be nice to have it as a command too.

**Comment by Rimmuru** [comment]
is it fine as is since thats only a command and no gui?
am i missing something?, as soon as i add a new cpp file it starts throwing errors in multiple places? unsure if its related to cmake since ive never used it before.

File: src/util/troll.hpp (hunk 2 above)
Comment (Loko415) [reply src/util/troll.hpp:9]: very easy to do lmao

Comment (Loko415) [reply src/util/troll.hpp:9]: the commands system is 10/10 and its pretty easy to convert stuff into commands

Comment (Yimura) [reply src/util/troll.hpp:9]: @Loko415 you still have got a problem with posting stuff into multiple comments for something that could've been edited in.

**Comment by tupoy-ya** [comment]
> Am i missing something? As soon as i add a new cpp file it starts throwing errors in multiple places? unsure if its related to cmake since ive never used it before.

This usually happens when CMake doesn't auto regenerate project files, you need to manually re-run CMake to start compiling new files.

> Is it fine as is? Since that's only a command and no gui?

Some stuff you have added already exists as commands so there is no reason to duplicate it.

**Comment by Rimmuru** [comment]
> > Am i missing something? As soon as i add a new cpp file it starts throwing errors in multiple places? unsure if its related to cmake since ive never used it before.
> 
> This usually happens when CMake doesn't auto regenerate project files, you need to manually re-run CMake to start compiling new files.
> 
> > Is it fine as is? Since that's only a command and no gui?
> 
> Some stuff you have added already exists as commands so there is no reason to duplicate it.

got it thank you

**Comment by Rimmuru** [comment]
hey guys, working on some replacement features, but i have a small issue. i dont actually know what to pass for the param of GET_ALL_VEHICLES native. any help?

**Comment by Aure7138** [comment]
> hey guys, working on some replacement features, but i have a small issue. i dont actually know what to pass for the param of GET_ALL_VEHICLES native. any help?

```cpp
auto get_all_vehicles = []() -> std::vector<Vehicle>
{
	std::vector<Vehicle> result;
	Vehicle vehs[300 * 2 + 2];
	vehs[0] = 300;
	int nums = VEHICLE::GET_ALL_VEHICLES(vehs);
	for (int i = 2; result.size() < nums && i < sizeof(vehs) / sizeof(vehs[0]); i++)
	{
		if (ENTITY::DOES_ENTITY_EXIST(vehs[i]) && !ENTITY::IS_ENTITY_DEAD(vehs[i], false))
		{
			result.push_back(vehs[i]);
		}
	}
	return result;
};
```

**Comment by Rimmuru** [comment]
> > hey guys, working on some replacement features, but i have a small issue. i dont actually know what to pass for the param of GET_ALL_VEHICLES native. any help?
> 
> ```c++
> auto get_all_vehicles = []() -> std::vector<Vehicle>
> {
> 	std::vector<Vehicle> result;
> 	Vehicle vehs[300 * 2 + 2];
> 	vehs[0] = 300;
> 	int nums = VEHICLE::GET_ALL_VEHICLES(vehs);
> 	for (int i = 2; result.size() < nums && i < sizeof(vehs) / sizeof(vehs[0]); i++)
> 	{
> 		if (ENTITY::DOES_ENTITY_EXIST(vehs[i]) && !ENTITY::IS_ENTITY_DEAD(vehs[i], false))
> 		{
> 			result.push_back(vehs[i]);
> 		}
> 	}
> 	return result;
> };
> ```

<3

**Comment by MarkEcza** [comment]
> hey guys, working on some replacement features, but i have a small issue. i dont actually know what to pass for the param of GET_ALL_VEHICLES native. any help?

You can also try using the CReplayInterface->CVehicleInterface->get_vehicle() function.

Small example would be;


```c++
rage::CReplayInterface* CReplayInterface_var = *g_pointers->m_replay_interface; //(YimMenu should already have this pointer)
for (int j = 0; j <= 300; j++) //max vehicle count is 300.
{
auto vehicle_ptr = CReplayInterface_var->m_vehicle_interface->get_vehicle(i);
     if (vehicle_ptr)
     {
        Vehicle vehicle_handle = g_pointers->m_ptr_to_handle(vehicle_ptr);

          //do stuff. 
     }

}
```

Although i haven't tested it. It should work.

**Comment by Yimura** [comment]
Is this ready?

@Rimmuru, I prefer it when a PR has a clear goal in mind that way I know when it's done. You've decided to add more from people that commented on this PR making it unclear when it was ready. Neither does your PR title make it clear what its goal was I had no idea if it was done or not.

Please mark your PR as a `draft` if you're still working on it and then mark it as ready when it is.

**Comment by Rimmuru** [comment]
guess idk how to properly fix the conflict /shrug

**Comment by Loko415** [comment]
memory leak on a plate. crashes the game every 5 seconds LMAO

**Comment by Rimmuru** [comment]
> memory leak on a plate. crashes the game every 5 seconds LMAO

ive been playing with it for hours. you arent gonna get a mem leak from using natives lmao

File: src/util/train.hpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		int nums = VEHICLE::GET_ALL_VEHICLES(vehs);
+		for (int i = 1; result.size() < nums && i < sizeof(vehs) / sizeof(vehs[0]); i++)
+		{
+			if (ENTITY::DOES_ENTITY_EXIST(vehs[i]) && !ENTITY::IS_ENTITY_DEAD(vehs[i], false))
+			{
+				result.push_back(vehs[i]);
+			}
+		}
+		return result;
+	};
```
Comment (Yimura) [inline src/util/train.hpp]: Why is index[0] set to `300`?
Why does the iterator start at 1?

`std::vector::size()` is a method returns the amount of entries in a vector.

Try taking a look at how the replay interface works, you can easily iterate of all the entities that exist on your local machine and push them onto a vector.

File: src/views/self/view_mobile.cpp [hunk 4]
```diff
@@ -7,26 +7,42 @@ namespace big
...
-		
-		components::sub_title("Merryweather");
 
+		components::sub_title("Merryweather");
 		ImGui::Separator();
 
 		components::button("Request Ammo Drop", [] {
 			mobile::merry_weather::request_ammo_drop();
-		});
+			});
```
Comment (Yimura) [inline src/views/self/view_mobile.cpp]: Indent

File: src/views/self/view_mobile.cpp [hunk 5]
```diff
@@ -7,26 +7,42 @@ namespace big
...
 
 		components::button("Helicopter Pickup", [] {
 			mobile::merry_weather::request_helicopter_pickup();
-		});
+			});
```
Comment (Yimura) [inline src/views/self/view_mobile.cpp]: Indent

File: src/views/self/view_mobile.cpp [hunk 6]
```diff
@@ -7,26 +7,42 @@ namespace big
...
+
+		components::button("Request Backup Helicopter", [] {
+			mobile::merry_weather::request_backup_helicopter();
+			});
```
Comment (Yimura) [inline src/views/self/view_mobile.cpp]: Indent

File: src/views/self/view_mobile.cpp [hunk 7]
```diff
@@ -7,26 +7,42 @@ namespace big
...
+
+		components::button("Request Airstrike", [] {
+			mobile::merry_weather::request_airstrike();
+			});
```
Comment (Yimura) [inline src/views/self/view_mobile.cpp]: Indent

File: src/views/self/view_mobile.cpp [hunk 8]
```diff
@@ -7,26 +7,42 @@ namespace big
...
 			);
-		});
+			});
+
+		components::sub_title("CEO Abilities");
+		ImGui::Separator();
+
+		components::button("Bullshark Testosterone", [] {
+			mobile::ceo_abilities::request_bullshark_testosterone();
+			});
```
Comment (Yimura) [inline src/views/self/view_mobile.cpp]: Indent

File: src/views/world/view_train.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+#include "script.hpp"
+#include "natives.hpp"
+#include "pointers.hpp"
+#include "views/view.hpp"
+#include "util/train.hpp"
+#include "services/gta_data/gta_data_service.hpp"
+#include "services/model_preview/model_preview_service.hpp"
+#include "services/players/player_service.hpp"
+
+#include <imgui_internal.h>
```
Comment (Yimura) [inline src/views/world/view_train.cpp]: Are all  these includes required?

File: src/views/world/view_train.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+			{
+				train::hijack_train();
+			});
+
+		ImGui::SameLine();
+		
+		components::button("Delete Train", []
+			{
+				train::delete_train();
+			});
```
Comment (Yimura) [inline src/views/world/view_train.cpp:21]: Indent

File: src/views/world/view_train.cpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+
+namespace big
+{
+	void view::train()
+	{
+		components::button("Hijack Train", []
+			{
+				train::hijack_train();
+			});
```
Comment (Yimura) [inline src/views/world/view_train.cpp:14]: Indent

File: src/views/world/view_train.cpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+		if (ImGui::IsItemHovered())
+			ImGui::SetTooltip("You cant delete the train while in it.");
+		
+		ImGui::SameLine();
+		
+		components::button("Exit Train", []
+			{
+				train::exit_train();
+			});
```
Comment (Yimura) [inline src/views/world/view_train.cpp:31]: Indent

**Review by Yimura — CHANGES_REQUESTED** [review]

**Comment by urmommine** [comment]
can u can u give me the code for teleport all to vespucci beach?

File: src/util/train.hpp (hunk 3 above)
Comment (Rimmuru) [reply src/util/train.hpp]: that function was written by one the guys above

Comment (Yimura) [reply src/util/train.hpp]: I'll take a look at replacing it with the replay interface.

Comment (Rimmuru) [reply src/util/train.hpp]: better?

**Review by Yimura — APPROVED** [review]

**Comment by Yimura** [comment]
Looks good to me, we can swap the replay interface later by directly using the entity pools, however this is a thing to be added in a different PR.

File: src/util/train.hpp [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		auto train = get_closest_train();
+
+		if (train != 0)
+		{
+			NETWORK::NETWORK_REQUEST_CONTROL_OF_ENTITY(train);
+
+			while (!NETWORK::NETWORK_HAS_CONTROL_OF_ENTITY(train))
+			{
+				script::get_current()->yield();
+			}
```
Comment (Yimura) [inline src/util/train.hpp]: Could end up infinitely looping causing the fiber pool to lockup, make use of the `entity::request_control_of()` function which has a built in breakout if it fails to get control of the entity.

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/util/train.hpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+#include "natives.hpp"
+#include "script.hpp"
+
+namespace big::train
+{
+	inline auto get_all_vehicles()
+	{
+		std::vector<Vehicle> result;
+		rage::CReplayInterface* CReplayInterface_var = *g_pointers->m_replay_interface; 
+		for (int i = 0; i <= 300; i++)
```
Comment (Aure7138) [inline src/util/train.hpp]: < 300 would be better, <= 300 is actually 301 cars

File: src/core/globals.hpp [hunk 15]
```diff
@@ -387,6 +387,12 @@ namespace big
 			NLOHMANN_DEFINE_TYPE_INTRUSIVE(spawn_ped, preview_ped)
 		} spawn_ped{};
 
+		struct train
+		{
+			bool derail_train = false;
+			bool drive_train = false;
+		} train{};
```
Comment (Aure7138) [inline src/core/globals.hpp:394]: Just a reminder, you can use NLOHMANN_DEFINE_TYPE_INTRUSIVE to save settings, the code looks good to me

Comment (Rimmuru) [reply src/core/globals.hpp:394]: the idea with them though was they wouldnt be saved

Comment (Aure7138) [reply src/core/globals.hpp:394]: ah, alright

File: src/util/train.hpp (hunk 14 above)
Comment (Rimmuru) [reply src/util/train.hpp]: agreed, sorted

File: src/util/train.hpp (hunk 13 above)
Comment (Rimmuru) [reply src/util/train.hpp]: assuming you meant entity::take_control_of, switched it over

File: src/core/globals.hpp (hunk 15 above)
Comment (Rimmuru) [reply src/core/globals.hpp:394]: :)

File: src/util/train.hpp (hunk 13 above)
Comment (Yimura) [reply src/util/train.hpp]: Yup, seems correct now.

**Review by Yimura — APPROVED** [review]

**Pull request merged**